[kube-prometheus-stack] add serviceName option to prometheus CR

trouaux opened the pull request:
<!--
Thank you for contributing to prometheus-community/helm-charts.
Before you submit this pull request we'd like to make sure you are aware of our technical requirements and best practices:

* https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#technical-requirements
* https://helm.sh/docs/chart_best_practices/

For a quick overview across what we will look at reviewing your PR, please read our review guidelines:

// TODO: add a REVIEW_GUIDELINES.md in prometheus-community/helm-charts
* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and help get your pull request merged quicker.

When updates to your pull request are requested, please add new commits and do not squash the history.
This will make it easier to identify new changes.
The pull request will be squashed anyways when it is merged.
Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them.
Once the pull request is opened, GitHub Actions will run across your changes and do some initial checks and linting.
These checks run very quickly.
Please check the results.
If you are contributing to this repository for the first time, a maintainer will need to approve those checks to run.
They are automatically requested as reviewers and will approve the workflows or ask you for changes once they get to it.

We would like these checks to pass before we even continue reviewing your changes.
-->

<!-- markdownlint-disable-next-line first-line-heading -->
#### What this PR does / why we need it
Adding serviceName option to Prometheus CR. This allows prometheus-operator reconciler to associate templated Service to Prometheus StatefulSet instead of creating an operated service.

see https://prometheus-operator.dev/docs/api-reference/api/#monitoring.coreos.com/v1.PrometheusSpec

#### Which issue this PR fixes

- fixes #5272 

#### Special notes for your reviewer

#### Checklist

<!-- [Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->
- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

jkroepke reviewed (changes requested):
In general, I like this change. Currently, the service is just prometheus-operated or alertmanager-operated and the are different from all other names.

However, this is a breaking changes. A lot of people may depend against the operated service names (including the helm chart itself + private deployments)

In order to approve this PR, my requirements are:
* fix all references to the hardcoded service names within the helm chart, e.g. https://github.com/prometheus-community/helm-charts/blob/f4f38ef92e6cb01433ac0e2546bed0774e3b9d7a/charts/kube-prometheus-stack/templates/grafana/configmaps-datasources.yaml#L54

* Ensure, that the service generated by helm differs from the service name generated by the operator (https://github.com/prometheus-community/helm-charts/blob/f4f38ef92e6cb01433ac0e2546bed0774e3b9d7a/charts/kube-prometheus-stack/templates/prometheus/service.yaml)
* Add a toggle to disable the service generated by helm
* Bump the major version
* Add a note into the README and add a snippet how to restore the orginal behavior

And what about Alertmanager?

trouaux commented:
> * fix all references to the hardcoded service names within the helm chart, e.g. https://github.com/prometheus-community/helm-charts/blob/f4f38ef92e6cb01433ac0e2546bed0774e3b9d7a/charts/kube-prometheus-stack/templates/grafana/configmaps-datasources.yaml#L54
> * Ensure, that the service generated by helm differs from the service name generated by the operator (https://github.com/prometheus-community/helm-charts/blob/f4f38ef92e6cb01433ac0e2546bed0774e3b9d7a/charts/kube-prometheus-stack/templates/prometheus/service.yaml)
> * Add a toggle to disable the service generated by helm
> * Bump the major version
> * Add a note into the README and add a snippet how to restore the orginal behavior

agree! I will work on those changes, thanks for the feedback

> And what about Alertmanager?

Currently alertmanager spec does not support serviceName option, but there is open issue for it and ThanosRuler. 
thanos-ruler-operated would be a third service

jkroepke commented:
> but there is open issue for it and ThanosRuler

good catch, but it seems implemented.

https://github.com/prometheus-operator/prometheus-operator/pull/5632

trouaux commented:
> > but there is open issue for it and ThanosRuler
> 
> good catch, but it seems implemented.
> 
> [prometheus-operator/prometheus-operator#5632](https://github.com/prometheus-operator/prometheus-operator/pull/5632)

this should be just passing the name of the operated service which is created by default right now `governingServiceName` 

definitely a place to think about when actually implementing the serviceName option

jkroepke commented:
Maybe keep this on hold until upstream changes are done.

I would like to avoid 2 breaking changes.

jkroepke commented:
Correct Thanos PR:

* https://github.com/prometheus-operator/prometheus-operator/pull/7325

jkroepke reviewed (changes requested):
So manually extent CRDs ist still a nogo - please remove them. If there are not present in upstream, we have to wait

trouaux commented:
> So manually extent CRDs ist still a nogo - please remove them. If there are not present in upstream, we have to wait

thanks for the review, updated crds with script and fixed some linting

jkroepke commented on charts/kube-prometheus-stack/Chart.yaml:
```diff
@@ -31,8 +31,8 @@ name: kube-prometheus-stack
 sources:
   - https://github.com/prometheus-community/helm-charts
   - https://github.com/prometheus-operator/kube-prometheus
-version: 69.8.2
```
please bump the major version and leave a note in README.md about this. This is necessary for each CRD changes.

jkroepke commented on charts/kube-prometheus-stack/UPGRADE.md:
```diff
@@ -1,5 +1,24 @@
+
+This version upgrades Prometheus-Operator to v0.81.0 now supporting serviceName for Prometheus/AM/ThanosRuler CRDs. If empty there is no changes to the creation of an operated service. If defined operated service will not be created and new serviceName will be used as governing service.
+
+Run these commands to update the CRDs before applying the upgrade.
```
````suggestion
Since [68.4.0](https://github.com/prometheus-community/helm-charts/pull/5175) it is also possible to use `crds.upgradeJob.enabled` for upgrading the CRDs.
For traditional upgrades, please run these commands to update the CRDs before applying the upgrade.
````

jkroepke commented:
LGTM to me.
